### PR TITLE
Improve custom data field usage

### PR DESCRIPTION
**Pull request by GermanGolota**

### CODE REVIEW

File: src/Aurora.Application/Models/SearchItem.cs [hunk 1]
```diff
@@ -1,5 +1,3 @@
 ﻿namespace Aurora.Application.Models;
 
-public record SearchItem<T> 
-    (ContentType ContentType, string ImagePreviewUrl, string SearchItemUrl, T? Data = null)
-    where T : SearchResultData;
\ No newline at end of file
+public record SearchItem(ContentType ContentType, string ImagePreviewUrl, string SearchItemUrl, SearchResultData? Data = null);
```
Comment (RuslanPr0g) [inline src/Aurora.Application/Models/SearchItem.cs:3]: null -> default?

File: src/Aurora.Application/Models/SearchResultData.cs [hunk 2]
```diff
@@ -1,5 +1,19 @@
-﻿namespace Aurora.Application.Models;
+﻿using System.Text.Json;
+using System.Text.Json.Serialization;
 
+namespace Aurora.Application.Models;
```
Comment (RuslanPr0g) [inline src/Aurora.Application/Models/SearchResultData.cs:4]: is it model or dto?

File: src/Aurora.Application/Models/SearchResultDto.cs [hunk 3]
```diff
@@ -2,8 +2,7 @@
 
 public class SearchResultDto
 {
-    //TODO: Add generic type preservation to SearchResultDto
-    public List<SearchItem<SearchResultData>>? Items { get; }
+    public List<SearchItem>? Items { get; }
```
Comment (RuslanPr0g) [inline src/Aurora.Application/Models/SearchResultDto.cs:5]: why list is nullable

File: src/Aurora.Application/Scrapers/ScraperRunner.cs [hunk 4]
```diff
@@ -24,7 +24,7 @@ public async Task<List<SearchResultDto>> Run(SearchRequestDto searchRequest, Fun
     {
         var options = searchRequest.Websites.Select(website => searchRequest.ContentTypes.Select(option => (website, option))).Flatten();
         var scrapers = await _collector.CollectFor(options);
-        IEnumerable<Task<(ValueOrNull<List<SearchItem<SearchResultData>>> result, IOptionScraper scraper)>> scrapingTasks = null!;
+        IEnumerable<Task<(ValueOrNull<List<SearchItem>> result, IOptionScraper scraper)>> scrapingTasks = null!;
```
Comment (RuslanPr0g) [inline src/Aurora.Application/Scrapers/ScraperRunner.cs:27]: why collection is null

File: src/Aurora.Scrapers/Services/PagingRunner.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        }
+        List<SearchItem> result = new();
+        for (int i = 0; i < maxPageNumber; i++)
+        {
+            bool failed = false;
+            try
+            {
+                var page = await loadPage(i, client);
+                await page.ResolveAsync(async document =>
+                {
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/Services/PagingRunner.cs:56]: hard to follow logic, but I guess it's okay considering our context, and still

File: src/Aurora.Scrapers/Services/PagingRunner.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                    return Task.CompletedTask;
+                });
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Failed to load page for '{scraperName}' with message '{msg}'", scraperName, e.Message);
+                failed = true;
+            }
+
+            if (failed || (options.UseLimitations && options.MaxItemsCount <= result.Count))
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/Services/PagingRunner.cs]: perhaps move to bool expression method to be more descriptive

File: src/Aurora.Scrapers/Total/FootFetishBooruTotalScraper.cs [hunk 7]
```diff
@@ -15,15 +15,15 @@ public FootFetishBooruTotalScraper(IHttpClientFactory clientFactory, IOptions<To
...
+    public async Task<IEnumerable<(List<string> Terms, List<SearchItem> Items)>> Scrap()
     {
         var config = _options.Value;
         var htmlDocument = new HtmlDocument
         {
             OptionFixNestedTags = true
         };
 
-        List<(List<string> Terms, List<SearchItem<SearchResultData>> Items)> result;
+        List<(List<string> Terms, List<SearchItem> Items)> result;
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/Total/FootFetishBooruTotalScraper.cs]: perhaps some value object (not value type) wrapper for this?

File: src/Aurora.Scrapers/Total/FootFetishBooruTotalScraper.cs [hunk 8]
```diff
@@ -38,7 +38,7 @@ public async Task<IEnumerable<(List<string> Terms, List<SearchItem<SearchResultD
                 {
                     pagesCount = Math.Min(config.MaxPagesCount, pagesCount);
                 }
-                Dictionary<List<string>, List<SearchItem<SearchResultData>>> items = new();
+                Dictionary<List<string>, List<SearchItem>> items = new();
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/Total/FootFetishBooruTotalScraper.cs]: here is the same problem, it feels like the lack of wrapper type

File: src/Aurora.Scrapers/Usings.cs [hunk 9]
```diff
@@ -6,4 +6,5 @@
 global using Aurora.Shared.Extensions;
 global using Aurora.Shared.Models;
 global using HtmlAgilityPack;
-global using Microsoft.Extensions.Options;
\ No newline at end of file
+global using Microsoft.Extensions.Options;
+global using Aurora.Scrapers.ResultData;
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/Usings.cs:10]: global using is good, approved

File: src/Aurora.Scrapers/ResultData/FootfetishBooruResultData.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Aurora.Scrapers.ResultData;
+
+public class FootfetishBooruResultData : SearchResultData
+{
+    public FootfetishBooruResultData(string[] tags)
+    {
+        Tags = tags;
+    }
+
+    public string[] Tags { get; set; }
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/ResultData/FootfetishBooruResultData.cs]: why array

File: src/Aurora.Scrapers/ResultData/FootfetishBooruResultData.cs [hunk 11]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Aurora.Scrapers.ResultData;
+
+public class FootfetishBooruResultData : SearchResultData
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/ResultData/FootfetishBooruResultData.cs]: it could be sealed record

File: src/Aurora.Scrapers/Total/FootFetishBooruTotalScraper.cs (hunk 7 above)
Comment (RuslanPr0g) [reply src/Aurora.Scrapers/Total/FootFetishBooruTotalScraper.cs]: this class can be sealed record though

File: src/Aurora.Shared/Models/ValueOrNull.cs [hunk 12]
```diff
@@ -34,6 +34,20 @@ public TResult Resolve<TResult>(Func<T, TResult> onValue, Func<string, TResult>
         return result;
     }
 
+    public Task ResolveAsync(Func<T, Task> onValue, Func<string, Task> onNull)
+    {
+        Task result;
+        if (HasValue)
```
Comment (RuslanPr0g) [inline src/Aurora.Shared/Models/ValueOrNull.cs]: use result.Then (you can add Then method like js has)

File: src/Aurora.Shared/Models/ValueOrNull.cs [hunk 13]
```diff
@@ -34,6 +34,20 @@ public TResult Resolve<TResult>(Func<T, TResult> onValue, Func<string, TResult>
...
+        {
+            result = onValue(Value!);
+        }
+        else
+        {
+            result = onNull.Invoke(NullMessage ?? "");
+        }
+        return result;
```
Comment (RuslanPr0g) [inline src/Aurora.Shared/Models/ValueOrNull.cs]: and here could be returned just.Then and it looks more clear on my magazines

File: src/Aurora.Shared/Models/ValueOrNull.cs [hunk 14]
```diff
@@ -79,12 +93,28 @@ public static ValueOrNull<T> CreateNull(string? nullMessage = null)
...
+    public static ValueOrNull<U> PipeValue<T, U>(this ValueOrNull<T> valueOrNull, Func<T, U> transform) =>
+        valueOrNull.Resolve(
+            value => transform(value).ToResult(), 
+            errorMessage => errorMessage.ToErrorResult<U>()
+            );
+
+    public static ValueOrNull<U> PipeValue<T, U>(this ValueOrNull<T> valueOrNull, Func<T, ValueOrNull<U>> transform) =>
+        valueOrNull.Resolve(
+            value => transform(value),
+            errorMessage => errorMessage.ToErrorResult<U>()
```
Comment (RuslanPr0g) [inline src/Aurora.Shared/Models/ValueOrNull.cs]: may be use not one error but list of errors, or add both, also you can add Error abstract class and Public and Private errors, such that Public is for user interface and Private just system-internal, this error and errors could get Error and ErrorList accordingly

File: tests/Aurora.Shared.Tests/Base/Child.cs [hunk 15]
```diff
@@ -1,12 +1,3 @@
...
-{
-    public Child(int parentProp, int childProp) : base(parentProp)
-    {
-        ParentProp = parentProp;
-        ChildProp = childProp;
-    }
-
-    public int ChildProp { get; }
-}
+public record Child(int ChildProp, int ParentProp) : Parent(ParentProp);
```
Comment (RuslanPr0g) [inline tests/Aurora.Shared.Tests/Base/Child.cs]: sealed

File: tests/Aurora.Shared.Tests/Base/Parent.cs [hunk 16]
```diff
@@ -1,11 +1,3 @@
...
-public class Parent
-{
-    public Parent(int parentProp)
-    {
-        ParentProp = parentProp;
-    }
-
-    public int ParentProp { get; set; }
-}
+public abstract record Parent(int ParentProp);
```
Comment (RuslanPr0g) [inline tests/Aurora.Shared.Tests/Base/Parent.cs]: Maybe add IdentityVersioned thing ???? this could describe Version and Audit (who and when modified), probably useful, but maybe not

File: tests/Aurora.Shared.Tests/Extensions/ValueConversionExtensionsTests.cs [hunk 17]
```diff
@@ -29,7 +29,7 @@ public void GetData_ShouldPreserveType_WhenChildTypeIsUsed()
         var jobj = obj.ToJObject();
         var str = jobj.ConvertToString();
         var jobj2 = str.ParseNullableJson();
-        var result = jobj2.ToData<Child>();
+        var result = jobj2.ToData<Parent>();
```
Comment (RuslanPr0g) [inline tests/Aurora.Shared.Tests/Extensions/ValueConversionExtensionsTests.cs:32]: we have tests?

**Review by RuslanPr0g — APPROVED** [review]

File: src/Aurora.Application/Models/SearchResultDto.cs (hunk 3 above)
Comment (GermanGolota) [reply src/Aurora.Application/Models/SearchResultDto.cs:5]: Because it can be set to null. In cases that we do not have results yet

File: tests/Aurora.Shared.Tests/Extensions/ValueConversionExtensionsTests.cs (hunk 17 above)
Comment (GermanGolota) [reply tests/Aurora.Shared.Tests/Extensions/ValueConversionExtensionsTests.cs:32]: Yes, we do

File: src/Aurora.Scrapers/Services/PagingRunner.cs (hunk 6 above)
Comment (GermanGolota) [reply src/Aurora.Scrapers/Services/PagingRunner.cs]: Moved

File: src/Aurora.Scrapers/Usings.cs (hunk 9 above)
Comment (GermanGolota) [reply src/Aurora.Scrapers/Usings.cs:10]: thanks!

File: src/Aurora.Shared/Models/ValueOrNull.cs (hunk 12 above)
Comment (GermanGolota) [reply src/Aurora.Shared/Models/ValueOrNull.cs]: introduced

File: src/Aurora.Shared/Models/ValueOrNull.cs (hunk 13 above)
Comment (GermanGolota) [reply src/Aurora.Shared/Models/ValueOrNull.cs]: introduced

File: src/Aurora.Shared/Models/ValueOrNull.cs (hunk 14 above)
Comment (GermanGolota) [reply src/Aurora.Shared/Models/ValueOrNull.cs]: That is a good suggestion, but we currently do not expose any errors to the end-user, so that would not be valuable at the moment

File: tests/Aurora.Shared.Tests/Base/Parent.cs (hunk 16 above)
Comment (GermanGolota) [reply tests/Aurora.Shared.Tests/Base/Parent.cs]: This is a test class.

File: tests/Aurora.Shared.Tests/Base/Child.cs (hunk 15 above)
Comment (GermanGolota) [reply tests/Aurora.Shared.Tests/Base/Child.cs]: Done

File: src/Aurora.Scrapers/ResultData/FootfetishBooruResultData.cs (hunk 11 above)
Comment (GermanGolota) [reply src/Aurora.Scrapers/ResultData/FootfetishBooruResultData.cs]: Good suggestion - implemented

File: src/Aurora.Scrapers/ResultData/FootfetishBooruResultData.cs (hunk 10 above)
Comment (GermanGolota) [reply src/Aurora.Scrapers/ResultData/FootfetishBooruResultData.cs]: We want to use this class as a value-object, so it would not make sense to have it be IEnumerable. Alternatively, we do not plan on adding anything to it, so it would not make sense to have it be a List.

File: src/Aurora.Application/Scrapers/ScraperRunner.cs (hunk 4 above)
Comment (GermanGolota) [reply src/Aurora.Application/Scrapers/ScraperRunner.cs:27]: We want to use those results it the catch block of the next try statement, so we need to assign some value to it.

Comment (GermanGolota) [reply src/Aurora.Application/Scrapers/ScraperRunner.cs:27]: It is needed to basically satisfy the compiler

File: src/Aurora.Application/Models/SearchItem.cs (hunk 1 above)
Comment (GermanGolota) [reply src/Aurora.Application/Models/SearchItem.cs:3]: That would not make sense in this context. As we want to have something like an Option<SearchResultData>, so we either have Nothing or Just SearchResultData

File: src/Aurora.Scrapers/Services/PagingRunner.cs (hunk 5 above)
Comment (GermanGolota) [reply src/Aurora.Scrapers/Services/PagingRunner.cs:56]: Refactored for readability

File: src/Aurora.Application/Models/SearchResultData.cs (hunk 2 above)
Comment (GermanGolota) [reply src/Aurora.Application/Models/SearchResultData.cs:4]: We use those terms interchangeably within the the context of this application
